diff --git a/llvm/include/llvm/TargetParser/Triple.h b/llvm/include/llvm/TargetParser/Triple.h
index 49ec8de..f15c0fd 100644
--- a/llvm/include/llvm/TargetParser/Triple.h
+++ b/llvm/include/llvm/TargetParser/Triple.h
@@ -204,6 +204,7 @@ public:
     ZOS,
     Haiku,
     RTEMS,
+    Ringos,
     NaCl,       // Native Client
     AIX,
     CUDA,       // NVIDIA CUDA
@@ -600,6 +601,11 @@ public:
     return getOS() == Triple::Win32;
   }
 
+  /// Tests whether the OS is RingOS.
+  bool isOSRingos() const {
+    return getOS() == Triple::Ringos;
+  }
+
   /// Checks if the environment is MSVC.
   bool isKnownWindowsMSVCEnvironment() const {
     return isOSWindows() && getEnvironment() == Triple::MSVC;
diff --git a/llvm/lib/TargetParser/Triple.cpp b/llvm/lib/TargetParser/Triple.cpp
index 0bbe8a3..4a334d1 100644
--- a/llvm/lib/TargetParser/Triple.cpp
+++ b/llvm/lib/TargetParser/Triple.cpp
@@ -264,6 +264,7 @@ StringRef Triple::getOSTypeName(OSType Kind) {
   case PS4: return "ps4";
   case PS5: return "ps5";
   case RTEMS: return "rtems";
+  case Ringos: return "ringos";
   case Solaris: return "solaris";
   case Serenity: return "serenity";
   case TvOS: return "tvos";
@@ -625,6 +626,7 @@ static Triple::OSType parseOS(StringRef OSName) {
     .StartsWith("zos", Triple::ZOS)
     .StartsWith("haiku", Triple::Haiku)
     .StartsWith("rtems", Triple::RTEMS)
+    .StartsWith("ringos", Triple::Ringos)
     .StartsWith("nacl", Triple::NaCl)
     .StartsWith("aix", Triple::AIX)
     .StartsWith("cuda", Triple::CUDA)
@@ -837,6 +839,7 @@ static Triple::ObjectFormatType getDefaultFormat(const Triple &T) {
   case Triple::x86_64:
     switch (T.getOS()) {
     case Triple::Win32:
+    case Triple::Ringos:
     case Triple::UEFI:
       return Triple::COFF;
     default:
diff --git a/llvm/unittests/TargetParser/TripleTest.cpp b/llvm/unittests/TargetParser/TripleTest.cpp
index 4db54a0..31b64df 100644
--- a/llvm/unittests/TargetParser/TripleTest.cpp
+++ b/llvm/unittests/TargetParser/TripleTest.cpp
@@ -163,6 +163,18 @@ TEST(TripleTest, ParsedIDs) {
   EXPECT_EQ(Triple::UnknownOS, T.getOS());
   EXPECT_EQ(Triple::UnknownEnvironment, T.getEnvironment());
 
+  T = Triple("x86_64-unknown-ringos-msvc");
+  EXPECT_EQ(Triple::x86_64, T.getArch());
+  EXPECT_EQ(Triple::UnknownVendor, T.getVendor());
+  EXPECT_EQ(Triple::Ringos, T.getOS());
+  EXPECT_EQ(Triple::MSVC, T.getEnvironment());
+
+  T = Triple("aarch64-unknown-ringos-msvc");
+  EXPECT_EQ(Triple::aarch64, T.getArch());
+  EXPECT_EQ(Triple::UnknownVendor, T.getVendor());
+  EXPECT_EQ(Triple::Ringos, T.getOS());
+  EXPECT_EQ(Triple::MSVC, T.getEnvironment());
+
   T = Triple("powerpcspe-unknown-freebsd");
   EXPECT_EQ(Triple::ppc, T.getArch());
   EXPECT_EQ(Triple::PPCSubArch_spe, T.getSubArch());
