diff --git a/clang/lib/CodeGen/CodeGenModule.cpp b/clang/lib/CodeGen/CodeGenModule.cpp
--- a/clang/lib/CodeGen/CodeGenModule.cpp
+++ b/clang/lib/CodeGen/CodeGenModule.cpp
@@ -142,7 +142,8 @@ createTargetCodeGenInfo(CodeGenModule &CGM) {
     AArch64ABIKind Kind = AArch64ABIKind::AAPCS;
     if (Target.getABI() == "darwinpcs")
       Kind = AArch64ABIKind::DarwinPCS;
-    else if (Triple.isOSWindows())
+    else if (Triple.isOSWindows() ||
+             (Triple.isOSRingos() && Triple.getEnvironment() == llvm::Triple::MSVC))
       return createWindowsAArch64TargetCodeGenInfo(CGM, AArch64ABIKind::Win64);
 
     return createAArch64TargetCodeGenInfo(CGM, Kind);
diff --git a/clang/test/CodeGen/aarch64-varargs-ringos-msvc.c b/clang/test/CodeGen/aarch64-varargs-ringos-msvc.c
new file mode 100644
--- /dev/null
+++ b/clang/test/CodeGen/aarch64-varargs-ringos-msvc.c
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -triple aarch64-unknown-ringos-msvc -emit-llvm -o - %s | FileCheck %s
+
+#include <stdarg.h>
+
+int simple_int(va_list ap) {
+// CHECK-LABEL: define dso_local i32 @simple_int
+  return va_arg(ap, int);
+// CHECK: [[RESULT:%[a-z_0-9]+]] = load i32, ptr %argp.cur
+// CHECK: ret i32 [[RESULT]]
+}
